firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java

```diff
@@ -0,0 +1,152 @@
+import androidx.annotation.VisibleForTesting;
+import com.google.firebase.firestore.index.IndexEntry;
+import com.google.firebase.firestore.util.AsyncQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implements the steps for backfilling indexes.
+ */
+public class IndexBackfiller {
+  /** How long we wait to try running index backfill after SDK initialization. */
+  private static final long INITIAL_BACKFILL_DELAY_MS = TimeUnit.SECONDS.toMillis(15);
+  /** Minimum amount of time between backfill checks, after the first one. */
+  private static final long REGULAR_BACKFILL_DELAY_MS = TimeUnit.MINUTES.toMillis(1);
+
+  private final SQLitePersistence persistence;
+
+  public IndexBackfiller(SQLitePersistence sqLitePersistence) {
+    this.persistence = sqLitePersistence;
+  }
+
+  public static class Results {
+    private final boolean hasRun;
+
+    private final int entriesAdded;
+    private final int entriesRemoved;
+
+    static IndexBackfiller.Results DidNotRun() {
+      return new IndexBackfiller.Results(/* hasRun= */ false, 0, 0);
+    }
+
+    Results(boolean hasRun, int entriesAdded, int entriesRemoved) {
+      this.hasRun = hasRun;
+      this.entriesAdded = entriesAdded;
+      this.entriesRemoved = entriesRemoved;
+    }
+
+    public boolean hasRun() {
+      return hasRun;
+    }
+
+    public int getEntriesAdded() {
+      return entriesAdded;
+    }
+
+    public int getEntriesRemoved() {
+      return entriesRemoved;
+    }
+  }
+
+  public class BackfillScheduler implements Scheduler {
+    private final AsyncQueue asyncQueue;
+    private final LocalStore localStore;
+    private boolean hasRun = false;
+    @Nullable private AsyncQueue.DelayedTask backfillTask;
+
+    public BackfillScheduler(AsyncQueue asyncQueue, LocalStore localStore) {
+      this.asyncQueue = asyncQueue;
+      this.localStore = localStore;
+    }
+
+    @Override
+    public void start() {
+      scheduleBackfill();
+    }
+
+    @Override
+    public void stop() {
+      if (backfillTask != null) {
+        backfillTask.cancel();
+      }
+    }
+
+    private void scheduleBackfill() {
+      long delay = hasRun ? REGULAR_BACKFILL_DELAY_MS : INITIAL_BACKFILL_DELAY_MS;
+      backfillTask =
+          asyncQueue.enqueueAfterDelay(
+              AsyncQueue.TimerId.INDEX_BACKFILL,
+              delay,
+              () -> {
+                localStore.backfillIndexes(IndexBackfiller.this);
+                hasRun = true;
+                scheduleBackfill();
+              });
+    }
+  }
+
+  public BackfillScheduler newScheduler(AsyncQueue asyncQueue, LocalStore localStore) {
+    return new BackfillScheduler(asyncQueue, localStore);
+  }
+
+  // TODO: Figure out which index entries to backfill.
+  public Results backfill() {
+    int numIndexesWritten = 0;
+    int numIndexesRemoved = 0;
+    return new Results(/* hasRun= */ true, numIndexesWritten, numIndexesRemoved);
+  }
+
+  public void addIndexEntry(IndexEntry entry) {
```

Comment:
These methods would probably also be private if it was not for your tests. They should also be marked `@VisibleForTesting` if that is true.

All of the test-only methods should also be package-private.